### PR TITLE
feat(signal): add Signal channel support

### PR DESCRIPTION
**Pull request by zayfod**
## Summary

- Integrates [signal-cli](https://github.com/AsamK/signal-cli) daemon via HTTP JSON-RPC as a nanobot channel
- Supports DMs and group chats with open/allowlist access policies
- Markdown→Signal text style conversion
- Typing indicators
- Attachment handling
- Group message context buffering
- Automatic reconnect with exponential backoff

Originally based on #601.

## Testing

- Added `tests/channels/test_signal_channel.py` and `tests/channels/test_signal_markdown.py`

### CODE REVIEW

File: nanobot/channels/signal.py [hunk 1]
```diff
@@ -0,0 +1,1133 @@
...
+        self._remember_account_id_alias(self.config.phone_number)
+
+        # Rolling message buffer for group context (group_id -> deque of messages)
+        # Each message is a dict with: sender_name, sender_number, content, timestamp
+        self._group_buffers: dict[str, deque] = {}
+
+    async def start(self) -> None:
+        """Start the Signal channel and connect to signal-cli daemon."""
+        if not self.config.phone_number:
+            self.self.logger.error("Signal account not configured")
```
Comment (chengyongru) [inline nanobot/channels/signal.py]: looks like a bug

**Review by chengyongru — CHANGES_REQUESTED** [review]

File: nanobot/channels/signal.py (hunk 1 above)
Comment (zayfod) [reply nanobot/channels/signal.py]: Addressed with https://github.com/HKUDS/nanobot/pull/3852/commits/90ec98e7cb0f2c35c3120fdd1eaf9da2b47196c6

File: nanobot/channels/signal.py [hunk 2]
```diff
@@ -0,0 +1,1321 @@
...
+                params: dict[str, Any] = {"message": chunk}
+                if chunk_styles[i]:
+                    params["textStyle"] = chunk_styles[i]
+                params.update(recipient_params)
+                if msg.media and i == 0:
+                    params["attachments"] = msg.media
+
+                response = await self._send_request("send", params)
+
+                if "error" in response:
```
Comment (chengyongru) [inline nanobot/channels/signal.py:561]: `_send_http_request` converts all exceptions into `{"error": {...}}` dicts, so this `if` branch is the only place the ChannelManager retry mechanism could be triggered. Raising here would let failed sends be automatically retried:

```suggestion
                if "error" in response:
                    self.logger.error(f"Error sending Signal message: {response['error']}")
                    raise RuntimeError(f"signal-cli send failed: {response['error']}")
```

File: nanobot/channels/signal.py [hunk 3]
```diff
@@ -0,0 +1,1321 @@
...
+                chat_id=chat_id,
+                content=content,
+                media=media_paths,
+                metadata={
+                    "timestamp": timestamp,
+                    "sender_name": sender_name,
+                    "sender_number": sender_number,
+                    "is_group": is_group_message,
+                    "group_id": group_id,
+                },
```
Comment (chengyongru) [inline nanobot/channels/signal.py:802]: Without `is_dm=True` for DMs, the base channel's `_handle_message` treats all denied messages as group messages and silently drops them instead of generating a pairing code. Adding this lets unapproved DM users receive a pairing code through the standard flow:

```suggestion
            },
            is_dm=not is_group_message,
        )
```

File: nanobot/channels/signal.py [hunk 4]
```diff
@@ -0,0 +1,1321 @@
...
+
+        # Direct message
+        chat_id = sender_number
+        if not self.config.dm.enabled:
+            self.logger.debug(f"Ignoring DM from {sender_id} (DMs disabled)")
+            return False, chat_id
+        if self.config.dm.policy == "allowlist":
+            if not self._sender_matches_allowlist(sender_id, self.config.dm.allow_from):
+                self.logger.debug(f"Ignoring DM from {sender_id} (policy: {self.config.dm.policy})")
+                return False, chat_id
```
Comment (chengyongru) [inline nanobot/channels/signal.py:870]: When a DM is denied by the allowlist here, it returns early before ever reaching `_handle_message`, so the pairing code is never generated.

One approach: instead of returning `(False, chat_id)`, let denied DMs fall through to `_handle_message(is_dm=True)` so the base channel can respond with a pairing code (this is how Slack handles it). Combined with the `is_dm` suggestion above, this would give Signal DM users the full pairing experience.

File: nanobot/channels/signal.py [hunk 5]
```diff
@@ -0,0 +1,1321 @@
...
+                        self.logger.debug(f"SSE line received: {line[:200]}")
+
+                    # SSE format handling
+                    if isinstance(line, str):
+                        # Empty line signals end of event
+                        if not line or line == ":":
+                            if event_buffer:
+                                # Try to parse the accumulated data
+                                data_str = ""
+                                try:
```
Comment (chengyongru) [inline nanobot/channels/signal.py:613]: Minor nit: the SSE spec says multiple `data:` lines should be joined with `
`. Not an issue today since signal-cli sends single-line JSON, but this would be more correct:

```suggestion
                                    data_str = "\n".join(event_buffer)
```

**Review by chengyongru — COMMENTED** [review]
Really nice work on this PR! A few things I particularly liked:

- **Excellent test coverage** — ~1900 lines of tests for ~1300 lines of implementation. The UTF-16 offset and chunk redistribution edge-case tests are especially thorough.
- **Clean separation** of `_check_inbound_policy` and `_assemble_inbound_content` makes the receive flow easy to follow.
- **Thoughtful identifier normalization** — the `+` prefix handling, case-insensitive UUID matching, and pipe-joined composite support for allowlists is solid.
- **The markdown→Signal textStyle converter** is a clever piece of work — the three-phase approach with opaque runs to prevent nested processing is elegant.

I have a few suggestions around the DM pairing flow. The core issue is that denied DM senders currently get silently dropped instead of receiving a pairing code — Telegram has the same gap, so you're in good company 😄. Two small changes should be enough to wire it up.

A couple of minor nits as well.

File: nanobot/channels/signal.py (hunk 2 above)
Comment (zayfod) [reply nanobot/channels/signal.py:561]: Addressed with https://github.com/HKUDS/nanobot/pull/3852/commits/4a39e79b9cc7f1020afa88a4c6ced55f9fdee292

File: nanobot/channels/signal.py (hunk 3 above)
Comment (zayfod) [reply nanobot/channels/signal.py:802]: Addressed with https://github.com/HKUDS/nanobot/pull/3852/commits/6a194a60909e74ef8b160db78f3053f404a34fb8

File: nanobot/channels/signal.py (hunk 4 above)
Comment (zayfod) [reply nanobot/channels/signal.py:870]: Addressed with https://github.com/HKUDS/nanobot/pull/3852/commits/c9f52986f9936b9b63fffa0e134deef91c6fad6b and https://github.com/HKUDS/nanobot/pull/3852/commits/d6a67f553f594b348b3436eee20f79e77f813686

File: nanobot/channels/signal.py (hunk 5 above)
Comment (zayfod) [reply nanobot/channels/signal.py:613]: Addressed with https://github.com/HKUDS/nanobot/pull/3852/commits/3592515ae34046870004129c6537f028c2148783

**Comment by chengyongru** [comment]
@zayfod I've pushed a few follow-up commits on top of your PR. Here's a summary of what changed and what I'd like you to verify:

**What changed (commit 3aefdc2):**

1. **Overrode `_handle_message`** — Messages that pass `_check_inbound_policy` now publish directly to the bus, bypassing the base-class `is_allowed()` check. This prevents cross-policy leakage: previously, `is_allowed()` had no message-type context, so e.g. `group.policy="open"` could cause a DM sender denied by `dm.policy="allowlist"` to be incorrectly allowed through. Denied DMs still call `super()._handle_message()` to issue pairing codes.

2. **SSE `data:` space handling** — Strip one optional leading space after `data:` per the SSE spec.

3. **Log format** — Converted f-string logger calls to loguru lazy formatting for consistency with other channels.

4. **Added end-to-end tests** — Including a cross-policy scenario (`dm.policy="allowlist" + group.policy="open"`) that verifies denied DM senders still get a pairing code.

All 188 tests pass.

**What to test manually:**

1. **Open DM** — Send a DM with `dm.policy: "open"`. The bot should reply normally (no pairing code).

2. **Allowlist DM — denied sender** — Set `dm.policy: "allowlist"` with an empty `dm.allowFrom`. Send a DM from an unlisted number. You should receive a pairing code, not a silent drop.

3. **Allowlist DM — allowed sender** — Add your number to `dm.allowFrom`. Confirm the bot replies normally.

4. **Group with mention** — In a group with `group.policy: "open", requireMention: true`, @mention the bot. It should respond.

5. **Group without mention** — Same group, send a message without mentioning the bot. It should be ignored.

6. **Markdown rendering** — Send a message that triggers a reply with bold, italic, code, etc. Verify the text styles render correctly in the Signal client.

If everything checks out I'll merge right away. Thanks!

**Comment by zayfod** [comment]
@chengyongru everything works as expected.

1. Open DM — Send a DM with dm.policy: "open". The bot should reply normally (no pairing code).

```
2026-05-20 10:20:19 | INFO  | signal | Data message from ***: groupInfo=None, groupV2=None, keys=['timestamp', 'message', 'expiresInSeconds', 'isExpirationUpdate', 'viewOnce']
2026-05-20 10:20:20 | INFO  | signal | Signal typing start sent for ***
2026-05-20 10:20:20 | INFO  | - | Processing message from signal:***|***: test
2026-05-20 10:20:26 | INFO  | - | Response to signal:***|***: Test successful
```

2. Allowlist DM — denied sender — Set dm.policy: "allowlist" with an empty dm.allowFrom. Send a DM from an unlisted number. You should receive a pairing code, not a silent drop.

```
2026-05-20 10:27:32 | INFO  | signal | Data message from ***: groupInfo=None, groupV2=None, keys=['timestamp', 'message', 'expiresInSeconds', 'isExpirationUpdate', 'viewOnce']
2026-05-20 10:27:32 | WARNING | signal | allow_from is empty — all access denied
2026-05-20 10:27:32 | INFO  | - | Generated pairing code ZCSG-8XPV for ***|***@signal
2026-05-20 10:27:33 | INFO  | signal | Sent pairing code ZCSG-8XPV to sender ***|*** in chat ***
```

3. Allowlist DM — allowed sender — Add your number to dm.allowFrom. Confirm the bot replies normally.

```
2026-05-20 10:32:59 | INFO  | signal | Data message from ***: groupInfo=None, groupV2=None, keys=['timestamp', 'message', 'expiresInSeconds', 'isExpirationUpdate', 'viewOnce']
2026-05-20 10:32:59 | INFO  | signal | Signal typing start sent for ***
2026-05-20 10:32:59 | INFO  | - | Processing message from signal:***|***: test
2026-05-20 10:33:04 | INFO  | - | Response to signal:***|***: Test received. All systems operational.
```

4. Group with mention — In a group with group.policy: "open", requireMention: true, @mention the bot. It should respond.

```
2026-05-20 10:24:11 | INFO  | signal | Data message from ***: groupInfo={'groupId': '***', 'groupName': '***', 'revision': 0, 'type': 'DELIVER'}, groupV2=None, keys=['timestamp', 'message', 'expiresInSeconds', 'isExpirationUpdate', 'viewOnce', 'mentions', 'groupInfo']
2026-05-20 10:24:11 | INFO  | signal | Signal typing start sent for ***
2026-05-20 10:24:11 | INFO  | - | Processing message from signal:***|***: [Recent group messages for context:]
***: test
---
[***]: test
2026-05-20 10:24:17 | INFO  | - | Response to signal:***|***: Test received. Systems are fully operational.
```

5. Group without mention — Same group, send a message without mentioning the bot. It should be ignored.

```
2026-05-20 10:23:43 | INFO  | signal | Data message from ***: groupInfo={'groupId': '***', 'groupName': '***', 'revision': 0, 'type': 'DELIVER'}, groupV2=None, keys=['timestamp', 'message', 'expiresInSeconds', 'isExpirationUpdate', 'viewOnce', 'groupInfo']
2026-05-20 10:23:43 | INFO  | signal | Ignoring group message (require_mention: True)
```

6. Markdown rendering — Send a message that triggers a reply with bold, italic, code, etc. Verify the text styles render correctly in the Signal client.

```
2026-05-20 10:33:50 | INFO  | signal | Data message from ***: groupInfo=None, groupV2=None, keys=['timestamp', 'message', 'expiresInSeconds', 'isExpirationUpdate', 'viewOnce']
2026-05-20 10:33:50 | INFO  | signal | Signal typing start sent for ***
2026-05-20 10:33:50 | INFO  | - | Processing message from signal:***|***: send a message with markdown formatting for a test
2026-05-20 10:34:00 | INFO  | - | Response to signal:***|***: Certainly, Sir. Here is a demonstration of the supported markdown formatting:

**Bold Text** and *Italic Text*
~~Striket...
```

Screenshot_20260520-103451


7. Allowlist group message — denied sender — Set group.policy: "allowlist" with an empty group.allowFrom. Send a message from an unlisted number. It is silently dropped.

```
2026-05-20 10:28:40 | INFO  | signal | Data message from ***: groupInfo={'groupId': '***', 'groupName': '***', 'revision': 0, 'type': 'DELIVER'}, groupV2=None, keys=['timestamp', 'message', 'expiresInSeconds', 'isExpirationUpdate', 'viewOnce', 'groupInfo']
2026-05-20 10:28:40 | INFO  | signal | Ignoring group message from *** (policy: allowlist)
```

**Review by chengyongru — APPROVED** [review]
Thanks! LGTM!
